cassiopeia/solver/DistanceSolver.py

```diff
@@ -0,0 +1,147 @@
+            raise DistanceSolveError(
+                "Please specify a dissimilarity map or dissimilarity function"
+            )
+
+        super().__init__(character_matrix, meta_data, priors)
+
+        self.dissimilarity_map = dissimilarity_map
+        self.dissimilarity_function = dissimilarity_function
+
+        if self.dissimilarity_map is None:
+
+            N = self.character_matrix.shape[0]
+            dissimilarity_map = np.zeros((N, N))
+            for i in range(character_matrix.shape[0]):
+
+                ind1 = self.character_matrix.iloc[i, :].values
+
+                for j in range(i + 1, character_matrix.shape[0]):
+
+                    ind2 = self.character_matrix.iloc[j, :].values
+                    dissimilarity_map[i, j] = dissimilarity_map[j, i] = self.dissimilarity_function(ind1, ind2)
+
+            self.dissimilarity_map = pd.DataFrame(
+                dissimilarity_map,
+                index=self.character_matrix.index,
+                columns=self.character_matrix.index,
+            )
+
+    def solve(self):
+        """A general bottom-up distance-based solver routine.
+
+        The general solver routine proceeds by iteratively finding pairs of
+        sapmles to join together into a "cherry" and then reform the
+        dissimilarity matrix with respect to this new cherry. The implementation
+        of how to find cherries and update the dissimilarity map is left to
+        subclasses of DistanceSolver. The function by default updates the
+        self.tree instance variable.
+        """
+
+        N = self.character_matrix.shape[0]
+
+        identifier_to_sample = dict(zip([str(i) for i in range(N)], self.character_matrix.index))
+
+        # instantiate a dissimilarity map that can be updated as we join
+        # together nodes.
+        _dissimilarity_map = self.dissimilarity_map.copy()
+
+        # instantiate a tree where all samples appear as leaves.
+        tree = nx.DiGraph()
+        tree.add_nodes_from(self.character_matrix.index)
+
+        while N > 2:
+
+            i, j = self.find_cherry(_dissimilarity_map.values)
+
+            # get indices in the dissimilarity matrix to join
+            node_i, node_j = (
+                _dissimilarity_map.index[i],
+                _dissimilarity_map.index[j],
+            )
+
+            new_node_name = str(len(tree.nodes))
+            tree.add_node(new_node_name)
+            tree.add_edges_from([(new_node_name, node_i), (new_node_name, node_j)])
+
+            _dissimilarity_map = self.update_dissimilarity_map(
+                _dissimilarity_map, (node_i, node_j), new_node_name
+            )
+
+            N = _dissimilarity_map.shape[0]
+
+        new_node_name = len(tree.nodes)
+        tree.add_node(new_node_name)
+
+        tree.add_edges_from([(new_node_name, i) for i in _dissimilarity_map.index])
+
+        tree = nx.relabel_nodes(tree, identifier_to_sample)
```

Comment:
Is there a reason to use the named indices of the cm and then relabel them to the numerical indices instead of just directly using the numerical indices?

Reply:
I think this is no longer an issue in the current code - I actually just add all the samples by name directly to the graph. If this isn't scalable enough, I'll only add unique samples and add a post-processing routine afterwards.